mlrun/runtimes/function.py

```diff
@@ -1217,6 +1116,23 @@ def compile_function_config(function: RemoteRuntime, client_version: str = None)
     tag = function.metadata.tag
     handler = function.spec.function_handler
 
+    def get_secrets(keys):
+        secrets = {}
+        if inside_k8s:
+            secrets = function._get_k8s().get_project_secret_data(project, keys)
+        for key in keys:
+            if key in builder_env:
+                secrets[key] = builder_env.get(key)
+        return secrets
+
+    if function.spec.build.source:
+        set_source_archive(spec, function, get_secrets)
```

Comment:
Why passing a function and not just the secrets themselves ? it feels like over complicating things, inside the `set_source_archive` when you know which secrets you need, go and bring them from the project secrets

Reply:
each archive need different secrets, i dont know in the top func which will be needed, i also need to implement the merge of secrets and build_env once instead of in each if/else section

Reply:
I'm saying move the whole `get_secrets` logic inside `set_source_archive`